### DISCUSSION TITLE
Intercept association row insertion

### DISCUSSION THREAD

**Discussion by novag** [sqlalchemy/sqlalchemy#12057] [Usage Questions]
Hi,

I need to intercept the insertion of an association row to set a specific attribute. I do not have to read the attribute afterwards, which is one reason why I do not want to use association objects. Unfortunately, I have not found a way to intercept the insert statement of the association row in the events “before_insert” or “before_flush”. I can see the parent and child objects but not the association row itself. Does anyone know of an alternative option or have I overlooked something?

Thanks!

**Comment by zzzeek** [comment]
the specific insert of these rows does not have an ORM event associated with it.  you would need to use connection level events like [before_execute](https://docs.sqlalchemy.org/en/20/core/events.html#sqlalchemy.events.ConnectionEvents.before_execute), which you could theoretically coordinate with ORM level events.

However the bigger question I'd want to ask is what's the actual use case here.  There are many ways to intercept the objects themselves, as well as ways to add table-level defaults to the association table itself.  but if it's not an association row with additional data it's not clear what you actually need to accompllish

**Reply by novag** [reply]
Hello @zzzeek,

Thanks for your detailed answer. I will try the before_execute event.

In my case each SQLAlchemy session object contains an id that identifies the author of a certain action. This id should be stored in the association row. Since this is a affecting a larger number of association tables, I do not want to convert all of them to an association object.

**Comment by ForkyTheEditor** [comment]
Bumping this up after almost a year as I have run into the exact same issue and have been unable to find an elegant solution for now.

I too have a many-to-many relationship with an additional id representing the tenant in a multitenant system. For easy development, I am looking to automatically add the tenant_id for every insert in every table.

**Reply by zzzeek** [reply]
here's "add table-level defaults to the association table itself", likely the easiest way since this gets you exactly at the INSERT for that table, I think this is pretty straightforward

```py
from __future__ import annotations

from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import event
from sqlalchemy import ExecutionContext
from sqlalchemy import ForeignKey
from sqlalchemy import Integer
from sqlalchemy import Table
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column
from sqlalchemy.orm import relationship
from sqlalchemy.orm import Session


class Base(DeclarativeBase):
    pass


def _get_tenant_id(context: ExecutionContext):
    return context.connection.info["tenant_id"]


atob = Table(
    "atob",
    Base.metadata,
    Column("a_id", ForeignKey("a.id"), primary_key=True),
    Column("b_id", ForeignKey("b.id"), primary_key=True),
    Column("session_id", Integer, nullable=False, default=_get_tenant_id),
)


class A(Base):
    __tablename__ = "a"

    id: Mapped[int] = mapped_column(primary_key=True)
    data: Mapped[str]
    bs: Mapped[list[B]] = relationship("B", secondary=atob)


class B(Base):
    __tablename__ = "b"
    id: Mapped[int] = mapped_column(primary_key=True)
    data: Mapped[str]

# optional.  set an event to propagate tenant_id from the connection
# to the session when a flush happens.   or you can do this manually
@event.listens_for(Session, "before_flush")
def set_tenant_id(session, flush_ctx, obj):
    session.connection().info["tenant_id"] = session.info["tenant_id"]


e = create_engine("sqlite://", echo=True)
Base.metadata.create_all(e)

with Session(e, info={"tenant_id": 5}) as s:

    # if not using an event, add the tenant id manually.  do this within
    # the scope of a transaction (that is, need to re-do it after a commit
    # unless sharing a connection)
    # s.connection().info["tenant_id"] = 5

    s.add(A(data="a1", bs=[B(data=f"b{i}") for i in range(10)]))
    s.commit()

```

**Reply by ForkyTheEditor** [reply]
The default function worked perfectly. Much simpler solution than my previous event trigger.
